### PR TITLE
fix: asyncio stalling inside notebooks

### PR DESCRIPTION
**Pull request by xianml**
tested in https://colab.research.google.com/drive/1NLFaFNxNxVgLPVgRevGv3NeyzJadNw96#scrollTo=AQX6pOz8BEu3&uniqifier=7

just a feasible solution. Any idea about other alternatives ?
@aarnphm 

image

### CODE REVIEW

File: openllm-core/src/openllm_core/utils/__init__.py [hunk 1]
```diff
@@ -71,7 +71,14 @@ def lenient_issubclass(cls: t.Any, class_or_tuple: type[t.Any] | tuple[type[t.An
...
   loop = asyncio.get_event_loop()
-  if loop.is_running(): return asyncio.run_coroutine_threadsafe(coro, loop).result()
+  if loop.is_running():
+    # in nootbook run_coroutine_threadsafe will got stuck for some reason (https://github.com/jupyter/notebook/issues/5261), so use asyncio.run instead
+    if in_notebook():
+      print('xxxxxx: RUNING in Ipython')
+      return asyncio.run(coro)
+    else:
+      print('xxxxxx: RUNING in Python')
+      return asyncio.run_coroutine_threadsafe(coro, loop).result()
```
Comment (aarnphm) [inline openllm-core/src/openllm_core/utils/__init__.py:81]:
```suggestion
    if in_notebook():
      return asyncio.run(coro)
    else:
      return asyncio.run_coroutine_threadsafe(coro, loop).result()
```

File: openllm-core/src/openllm_core/utils/__init__.py [hunk 2]
```diff
@@ -71,7 +71,12 @@ def lenient_issubclass(cls: t.Any, class_or_tuple: type[t.Any] | tuple[type[t.An
 
 def ensure_exec_coro(coro: t.Coroutine[t.Any, t.Any, t.Any]) -> t.Any:
   loop = asyncio.get_event_loop()
-  if loop.is_running(): return asyncio.run_coroutine_threadsafe(coro, loop).result()
+  if loop.is_running():
+    # in nootbook run_coroutine_threadsafe will got stuck for some reason (https://github.com/jupyter/notebook/issues/5261), so use asyncio.run instead
```
Comment (aarnphm) [inline openllm-core/src/openllm_core/utils/__init__.py:75]: This has to do with jupyter notebook has a running event loop

Comment (xianml) [reply openllm-core/src/openllm_core/utils/__init__.py]: yes.

File: openllm-core/src/openllm_core/utils/__init__.py [hunk 3]
```diff
@@ -70,6 +70,11 @@ def lenient_issubclass(cls: t.Any, class_or_tuple: type[t.Any] | tuple[type[t.An
     raise
 
 def ensure_exec_coro(coro: t.Coroutine[t.Any, t.Any, t.Any]) -> t.Any:
+  '''
+  in nootbook run_coroutine_threadsafe will got stuck for some reason (https://github.com/jupyter/notebook/issues/5261), so use asyncio.run instead
+  '''
+  if in_notebook():
+    return asyncio.run(coro)
```
Comment (aarnphm) [inline openllm-core/src/openllm_core/utils/__init__.py:77]:
```suggestion
  if in_notebook(): return asyncio.run(coro)  # For running coroutine in notebooks see https://github.com/jupyter/notebook/issues/5261 
```

**Review by aarnphm — APPROVED** [review]

**Pull request merged**